### ISSUE TITLE
RF upcast to f32 for certain ops (layernorm, attention, etc) (bf16/f16)

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1741]
RF doesn't do any special logic to upcast to f32 when the input is f16 or bf16 or so (e.g. via Torch AMP) for modules like `LayerNorm`. But maybe it should.

Of course, we should not just change the behavior. This should be an option. And maybe a new behavior version which changes the default of the option.

See also what others are doing:

Note, the behavior has changed in more recent versions of PyTorch for some of the PyTorch modules (e.g. `torch.nn.LayerNorm`, changed in 2.6 or so?).

Examples of LayerNorm, RMSNorm casting to f32:
* [HF Transformers Llama4](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama4/modeling_llama4.py)
* PyTorch itself in more recent versions.
* Torchtune.

Examples of rotary embedding (RoPE) casting to f32:
* [HF Transformers Llama4](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama4/modeling_llama4.py)

Examples of attention casting to f32:
* [Llama3](https://github.com/meta-llama/llama3/blob/main/llama/model.py), only for the softmax
 * PyTorch FlexAttention (I think...)

Examples of attention (intentionally) leaving the orig type:
* [HF Transformers Llama4](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama4/modeling_llama4.py) "llama4 doesn't cast attn weights to fp32"

Cross entropy / softmax on final logits: ...
